inject/maven-plugin/src/main/java/io/helidon/inject/maven/plugin/AbstractApplicationCreatorMojo.java

```diff
@@ -176,226 +269,250 @@ LinkedHashSet<Path> getModulepathElements() {
     }
 
     boolean hasModuleInfo() {
-        return getSourceRootPaths().stream()
-                .anyMatch(p -> new File(p.toFile(), REAL_MODULE_INFO_JAVA_NAME).exists());
+        return sourceRootPaths()
+                .stream()
+                .anyMatch(p -> Files.exists(p.resolve(ModuleInfo.FILE_NAME)));
+    }
+
+    protected Optional<Path> findModuleInfo(List<Path> sourcePaths) {
+        return sourcePaths.stream()
+                .map(it -> it.resolve(ModuleInfo.FILE_NAME))
+                .filter(Files::exists)
+                .findFirst();
     }
 
     /**
-     * Favors the 'test' module-info if available, and falls back to 'main' module-info.
+     * @return This represents the set of services that we already code-gen'ed
+     */
+    protected Set<TypeName> serviceTypeNamesForExclusion() {
+        return Set.of();
+    }
+
+    Set<TypeName> toNames(List<ServiceProvider<Object>> services) {
+        Map<TypeName, ServiceProvider<?>> result = new LinkedHashMap<>();
+        services.forEach(sp -> {
+            ServiceProvider<?> rootProvider = toRootProvider(sp);
+            TypeName serviceType = rootProvider.serviceType();
+            ServiceProvider<?> prev = result.put(serviceType, rootProvider);
+            if (prev != null) {
+                if (!(prev instanceof ServiceProviderProvider)) {
+                    throw new CodegenException("There are two registrations for the same service type: " + prev + " and " + sp);
+                }
+                getLog().debug("There are two registrations for the same service type: " + prev + " and " + sp);
+            }
+        });
+        return new TreeSet<>(result.keySet());
+    }
+
+    void warn(String msg) {
+        getLog().warn(msg);
+
+        if (failOnWarning()) {
+            throw new CodegenException(msg);
+        }
+    }
+
+    protected abstract CodegenScope scope();
+
+    /**
+     * Creates a new classloader.
      *
-     * @param location the location for the located module-info
-     * @return the module-info descriptor to return or null if none is available
+     * @param classPath the classpath to use
+     * @param parent    the parent loader
+     * @return the loader
      */
-    ModuleInfoDescriptor getAnyModuleInfo(AtomicReference<File> location) {
```

Comment:
How did you verify this refactor? Seems like its not computing the test application and the non-test application differently.

Reply:
It is, as the method to get the classpath and module are implemented differently in test mojo and production mojo

Reply:
Check line 124, where I use the current modules source root path (either test or production, depending on Mojo in use). Then on the next line I get the "production" source as a fallback to get exported packages.